File: smart-hub-direct/src/main/scala/io/github/tassiLuca/hub/application/ThermostatManager.scala
package io.github.tassiLuca.hub.application

import gears.async.{Async, AsyncOperations, ReadableChannel}
import io.github.tassiLuca.hub.core.ports.{AlertSystemComponent, DashboardServiceComponent, HeaterComponent}
import io.github.tassiLuca.hub.core.{
  SensorHealthCheckerComponent,
  TemperatureEntry,
  ThermostatComponent,
  ThermostatHourlyScheduler,
  ThermostatScheduler,
}
import io.github.tassiLuca.rears.{Controller, bufferWithin}

import scala.concurrent.duration.DurationInt
import scala.language.postfixOps

/** The thermostat manager. */
trait ThermostatManager
    extends ThermostatComponent[ThermostatHourlyScheduler]
    with SensorHealthCheckerComponent[TemperatureEntry]
    with HeaterComponent
    with AlertSystemComponent
    with DashboardServiceComponent:
  override val thermostat: Thermostat = Thermostat(ThermostatScheduler.byHour(19))
  override val sensorHealthChecker: SensorHealthChecker = SensorHealthChecker()
  private val samplingWindow = 10 seconds

  /** Runs the manager, spawning a new controller consuming the given [[source]] of events. */
  def run(source: ReadableChannel[TemperatureEntry])(using Async, AsyncOperations): Unit =
    thermostat.asRunnable.run
    sensorHealthChecker.asRunnable.run
    Controller.oneToMany(
      publisherChannel = source,
      consumers = Set(thermostat, sensorHealthChecker),
      transformation = r => r.bufferWithin(samplingWindow),
    ).run

File: smart-hub-direct-kt/src/main/kotlin/io/github/tassiLuca/hub/application/ThermostatManager.kt
package io.github.tassiLuca.hub.application

import io.github.tassiLuca.hub.core.SensorHealthChecker
import io.github.tassiLuca.hub.core.TemperatureEntry
import io.github.tassiLuca.hub.core.Thermostat
import io.github.tassiLuca.hub.core.ports.DashboardService
import kotlinx.coroutines.flow.Flow
import kotlin.coroutines.CoroutineContext
import kotlin.time.Duration.Companion.seconds

/** Manages the thermostat and the temperature sensors. */
class ThermostatManager(private val dashboardService: DashboardService, coroutineContext: CoroutineContext) {

    /** The thermostat. */
    val thermostat = Thermostat(TEMPERATURE_TARGET, SAMPLING_WINDOW, dashboardService, coroutineContext)

    /** The temperature sensors checker. */
    val temperatureSensorsChecker = SensorHealthChecker<TemperatureEntry>(
        SAMPLING_WINDOW,
        coroutineContext,
        dashboardService,
    )

    /** Runs the thermostat and the temperature sensors. */
    suspend fun run(sensorSource: Flow<TemperatureEntry>) {
        dashboardService.updateSchedule(mapOf("ALWAYS" to "09-24" to "19.0"))
        thermostat.run()
        temperatureSensorsChecker.run()
        sensorSource.collect {
            thermostat.react(it)
            temperatureSensorsChecker.react(it)
        }
    }

    companion object {
        private const val TEMPERATURE_TARGET = 19.0
        private val SAMPLING_WINDOW = 10.seconds
    }
}
